sdk/python/kfp/dsl/_container_op.py

```diff
@@ -12,309 +12,940 @@
+        warnings.warn(
+            '`dsl.ContainerOp.%s` will be removed in future releases. '
+            'Use `dsl.ContainerOp.container.%s` instead.' %
+            (op_name, container_name), PendingDeprecationWarning)
+        return func(*args, **kwargs)
 
-  def _validate_positive_number(self, str_value, param_name):
-    "Validate a given string is in positive integer format."
+    return _wrapped
 
-    try:
-      int_value = int(str_value)
-    except ValueError:
-      raise ValueError('Invalid {}. Should be integer.'.format(param_name))
 
-    if int_value <= 0:
-      raise ValueError('{} must be positive integer.'.format(param_name))
+def _create_getter_setter(prop):
+    """Create a tuple of getter and setter methods for a property in `Container`."""
+    def _getter(self):
+        return getattr(self._container, prop)   
+    def _setter(self, value):
+        return setattr(self._container, prop, value)
+    return _getter, _setter
 
-  def add_resource_limit(self, resource_name, value):
-    """Add the resource limit of the container.
 
-    Args:
-      resource_name: The name of the resource. It can be cpu, memory, etc.
-      value: The string value of the limit.
+def _proxy_container_op_props(cls: "ContainerOp"):
+    """Takes the `ContainerOp` class and proxy the PendingDeprecation properties
+    in `ContainerOp` to the `Container` instance.
     """
-
-    self.resource_limits[resource_name] = value
-    return self
-
-  def add_resource_request(self, resource_name, value):
-    """Add the resource request of the container.
-
-    Args:
-      resource_name: The name of the resource. It can be cpu, memory, etc.
-      value: The string value of the request.
-    """
-
-    self.resource_requests[resource_name] = value
-    return self
-
-  def set_memory_request(self, memory):
-    """Set memory request (minimum) for this operator.
-
-    Args:
-      memory: a string which can be a number or a number followed by one of
-              "E", "P", "T", "G", "M", "K".
+    # properties mapping to proxy: ContainerOps.<prop> => Container.<prop>
+    prop_map = dict(image='image', env_variables='env')
+    # itera and create class props
+    for op_prop, container_prop in prop_map.items():
+        # create getter and setter
+        _getter, _setter = _create_getter_setter(container_prop)
+        # decorate with deprecation warning
+        getter = deprecation_warning(_getter, op_prop, container_prop)
+        setter = deprecation_warning(_setter, op_prop, container_prop)
+        # update attribites with properties
+        setattr(cls, op_prop, property(getter, setter))
+    return cls
+
+
+def as_list(value: Any, if_none: Union[None, List] = None) -> List:
+    """Convert any value except None to a list if not already a list."""
+    if value is None:
+        return if_none
+    return value if isinstance(value, list) else [value]
+
+
+def create_and_append(current_list: Union[List[T], None], item: T) -> List[T]:
```

Comment:
If this is internally used in this module, could we rename it such that it starts with an underscore?